### DISCUSSION TITLE
Upserting Relay Paginated List

### DISCUSSION THREAD

**Discussion by elilabes** [urql-graphql/urql#3207] [Q&A]
👋 hello!

I have a query that has the following structure:

```graphql
query players($first: Int, $after: String) {
  me {
    profile {
      id
      players(first: $first, after: $after) {
        edges {
          node {
            id
            name
            avatarUrl
            userId
            inviteId
          }
        }
        pageInfo {
          startCursor
          endCursor
          hasNextPage
          hasPreviousPage
        }
      }
    }
  }
}
```

I use the built in `relayPagination` to turn the list of nodes into a single list.

```ts
export const cache = offlineExchange<GraphCacheConfig>({
  resolvers: {
    Profile: {
      players: relayPagination(),
    },
  },
});
```

I use this in my React components via GraphQL Codegen for Urql
```tsx
const [{ data  }] = usePlayersQuery({
     variables: {
       first: 10,
       after: cursor,
    },
 });
 
 console.log(data.me.profile.players.edges)
```

This works great, and I load more information on scroll (using react-native).

The problem I have, that I have been trying to resolve using the docs to no avail, is how I can then add to/update to the list from a mutation.

The mutation is structured like this.
```graphql
mutation upsertPlayer($input: UpsertPlayerInput!) {
  upsertPlayer(input: $input) {
    id
    name
    email
    inviteId
    avatarUrl
  }
}
```
This is where I am at the moment with the local mutation. It has a bug where it will add the new/updated player to every paginated query in the cache.

I could change it to look for existing players, then on the last item of playerData, if it hasn't found one then create it. That would work but I am wondering if there is a better way to do this?

While this code currently works for when there are less than 10 players (e.g. the `data` object get returned in the correct shape), it doesn't then show up in the  `usePlayersQuery` hook for some reason.

```ts
export const cache = offlineExchange<GraphCacheConfig>({
  resolvers: {
    Profile: {
      players: relayPagination(),
    },
  },
  updates: {
    Mutation: {
      upsertPlayer: (parent, _, cache) => {
        const profile = cache.keyOfEntity({
          __typename: "Profile",
          id: getUserId(),
        });

        const playersData = cache.inspectFields(profile).filter(field => field.fieldName === "players");
        for (const playerData of playersData) {
          cache.updateQuery({ query: PlayersDocument, variables: playerData.arguments }, (data: PlayersQuery | null) => {
            const newPlayer = parent.upsertPlayer as Player;
            const player = data?.me.profile.players.edges.find(player => player.node.id === newPlayer.id);
            if (!player) {
              data?.me.profile.players.edges.push({
                node: newPlayer,
              });

              return data;
            }

            player.node = newPlayer;

            return data;
          });
        }
      },
});
```

Appreciate any help - thanks 😄

**Comment by elilabes** [comment]
I changed the above slightly as I realised updated to existing data is already handled by the normalised cache.

```ts
// Find queries that are fetching `me.profile.players` records.
// There can be multiple, and they are split by their arguments
// which is our case our `first` and `after` used for pagination.
const meProfilePlayerQueries = cache.inspectFields(meProfile).filter(field => field.fieldName === "players");

meProfilePlayerQueries.forEach((query, index) => {
     cache.updateQuery({ query: PlayersDocument, variables: query.arguments }, (data: PlayersQuery | null) => {
         const player = parent.upsertPlayer as Player;
         const existingPlayer = data?.me.profile.players.edges.find(item => item.node.id === player.id);
         const isLastQuery = index === meProfilePlayerQueries.length - 1;
    
         // We only want to add new players to the last query.
         // If the player already exists, then the normalized cache
         // will handle updating the existing record.
         if (!existingPlayer && isLastQuery) {
            data?.me.profile.players.edges.push({
              __typename: "PlayerEdge",
              node: player,
            });
          }
    
          return data;
    });
  });
```

This works, but it doesn't feel very elegant.

Maybe I am just thinking about it wrong, but being able to get the full list of edges as though it was returned by the `relayPagination` would be useful as it would then allow sorting the new item into the right place. How that works with then mapping back to the individual query is where that gets difficult though.

**Reply by kitten** [reply]
So, essentially the problem is that there are several ways of doing pagination in `urql` (or similar GraphQL caches, or caches in general for that matter)

An alternative would be to instead render a list of fragmented components, i.e. each page gets a component. Often this isn't an option however in React Native, unless you're looking to jump through some hoops.

Another way would be to do it more similarly to what Apollo recommends by writing an updater and add it to `updates` that creates a "merged" field in the cache every time you get a result and adds to it and then have a resolver that returns that merged list on demand.

However, the `relayPagination` helper is a resolver that combines them on demand.
The point being, you can't get a list of edges and expect to be able to update them. After all, they have to be written somewhere.

However, there is a simple trick to how relay pagination works, straight copied from how Relay does it itself.

Each page that is stored is actually retrieved in order, and neither Relay nor Graphcache's `relayPagination` helper actually bothers sorting them, because they're expected to be retrieved in order.

This means that you can often cheat by:
- running `inspectFields`
- filtering out the fields you don't need
- picking the last one and updating only that one

You also often don't have to filter out existing items since `relayPagination` already deduplicates. Although it obviously doesn't hurt.

Also, keep in mind that you can update the list with as little as a fragment that only contains the list field, edge, and keyable fields of the node.
